### PR TITLE
Skip step AdamW

### PR DESCRIPTION
**Pull request by 2015aroras**
This PR implements `SkipStepAdamW`. It's not ideal since it doesn't support fused ops or similar at present. I changed default AdamW to also use the same underlying implementation

### CODE REVIEW

File: src/olmo_core/optim/adamw.py [hunk 1]
```diff
@@ -47,6 +48,129 @@ def adamw_step(
...
+
+    def __init__(
+        self,
+        params,
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 1e-2,
+        foreach: Optional[bool] = None,
+        fused: Optional[bool] = None,
```
Comment (dirkgr) [inline src/olmo_core/optim/adamw.py:64]: These two are ignored?

File: src/olmo_core/optim/adamw.py [hunk 2]
```diff
@@ -47,6 +48,129 @@ def adamw_step(
...
+        rolling_interval_length: int = 128,
+        sigma_factor: int = 6,
```
Comment (dirkgr) [inline src/olmo_core/optim/adamw.py:119]: Can you check whether this would have caught our spikes?

File: src/olmo_core/optim/adamw.py (hunk 1 above)
Comment (2015aroras) [reply src/olmo_core/optim/adamw.py:64]: Yes

File: src/olmo_core/optim/adamw.py [hunk 3]
```diff
@@ -61,5 +185,25 @@ class AdamWConfig(OptimConfig):  # NOTE: omagaconf doesn't like "OptimConfig[tor
...
+    """
+    Configuration class for building a :class:`SkipStepAdamW` optimizer.
+    """
+
+    lr: float = 1e-3
+    betas: Tuple[float, float] = (0.9, 0.999)
+    eps: float = 1e-8
+    weight_decay: float = 1e-2
+    foreach: Optional[bool] = None
+    fused: Optional[bool] = None
```
Comment (epwalsh) [inline src/olmo_core/optim/adamw.py:203]: These aren't used, so let's remove them.

File: src/olmo_core/optim/adamw.py [hunk 4]
```diff
@@ -61,5 +185,25 @@ class AdamWConfig(OptimConfig):  # NOTE: omagaconf doesn't like "OptimConfig[tor
     fused: Optional[bool] = None
 
     @classmethod
-    def optimizer(cls) -> Type[torch.optim.AdamW]:
-        return torch.optim.AdamW
```
Comment (epwalsh) [inline src/olmo_core/optim/adamw.py:65]: Why change this? We lose foreach and fused options unnecessarily.

**Review by epwalsh — CHANGES_REQUESTED** [review]
> It's not ideal since it doesn't support fused ops or similar at present. I changed default AdamW to also use the same underlying implementation.

I don't see a good reason for this. Let's default to the built-in AdamW so we can still use the `foreach` and `fused` options.

**Comment by 2015aroras** [comment]
At this point I've made so many other relevant changes to make this work that it might be easier to just merge the 32B when we're back

**Pull request closed without merging**